cbmc_viewer/memopt.py

```diff
@@ -0,0 +1,204 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""CBMC memory operation metric.
+
+This module displays calls to memory operations: memcpy, memcmp, memmove,
+memset, memchr, memccpy. It include SSA the function call expression,
+source location and total number of calls to these operations.
+Inputs to the module is a json file obtained by running CBMC with the
+--show-goto-functions parse option, the project source and reporting
+directories.
+"""
+
+import logging
+import json
+
+import voluptuous
+import voluptuous.humanize
+
+from cbmc_viewer import filet
+from cbmc_viewer import parse
+from cbmc_viewer import srcloct
+from cbmc_viewer import templates
+from cbmc_viewer import util
+
+JSON_TAG = "viewer-memop"
+
+################################################################
+
+VALID_MEMOP = voluptuous.schema_builder.Schema({
+    'funcCall': str,
+    'sourceLoc': srcloct.VALID_SRCLOC
+    },required=True)
+
+VALID_MEMOP_SUMMARY = voluptuous.schema_builder.Schema({
+    'summary': {
+        'list': [VALID_MEMOP],
+        'count': int}
+    },required=True)
+
+################################################################
+
+class MemOpSummary:
+
+    def __init__(self, json_file, srcdir):
+        self.summary = get_memop_metrics(json_file, srcdir)
+        self.validate()
+
+    def __repr__(self):
+        """A dict representation of memop summary."""
+
+        self.validate()
+        return self.__dict__
+
+    def __str__(self):
+        """A string representation of memop summary."""
+
+        return json.dumps({JSON_TAG: self.__repr__()}, indent=2)
+
+    def validate(self):
+        """Validate members of a summary object."""
+
+        return voluptuous.humanize.validate_with_humanized_errors(
+            self.__dict__, VALID_MEMOP_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir=None):
+        """Write memop metrics to a file or stdout."""
+
+        util.dump(self, filename, outdir)
+
+    def render_report(self, filename=None, outdir=None):
+        """Write the memop summary to a file rendered as html."""
+
+        memop_html = templates.render_memop_summary(self.summary)
+        util.dump(memop_html, filename or "memop.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_FUNCTIONS_KEY = 'functions'
+JSON_BODY_AVAIL_KEY = 'isBodyAvailable'
+JSON_INSTRUCTIONS_KEY = 'instructions'
+JSON_INSTRUCTION_ID_KEY = 'instructionId'
+JSON_OPERANDS_KEY = 'operands'
+JSON_FUNCTION_INSTR_KEY = 'instruction'
+JSON_SOURCE_LOC_KEY = 'sourceLocation'
+
+# Json keys for CBMC operands
+JSON_ID_KEY = 'id'
+JSON_NAMED_SUB_KEY = 'namedSub'
+JSON_IDENTIFIER_KEY = 'identifier'
+
+# List of memory operations
+list_mem_func = ["memcmp","memcpy","memmove","memset","memccpy","memchr"]
+
+# Example cbmc json output
+# [
```

Comment:
Is this json output format your invention or something defined by cbmc itself?

Reply:
This is defined by CBMC. Its the standard json output for `--show-goto-functions`

Reply:
This is the output of ---show-goto-functions.   This gives a list of functions.  For each function there is a list of instructions given by `instructions` tag. Every instruction has a `guard`, the `instruction` string giving the source location and the instruction itself separated by the newline delimiter, and `instruction_id` giving the instruction type and we are looking for the function calls, `operands` keeps track of the codet cbmc object stores the symbol of the function called and the arguments where `namedsub` -> `id` = symbol is the symbol for the function call and we extract the symbol name.    Only have the instructions tag if isBodyAvailable is true.